SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,528 @@
+/*
+ * Emitter.cs
+ * SnowplowTracker
+ * 
+ * Copyright (c) 2015 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ * 
+ * Authors: Joshua Beemster
+ * Copyright: Copyright (c) 2015 Snowplow Analytics Ltd
+ * License: Apache License Version 2.0
+ */
+
+using System;
+using System.Net;
+using System.Web;
+using System.IO;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Threading;
+using SnowplowTracker.Payloads;
+using SnowplowTracker.Enums;
+using SnowplowTracker.Requests;
+using SnowplowTracker.Storage;
+using SnowplowTracker.Collections;
+using UnityHTTP;
+
+namespace SnowplowTracker {
+	public class Emitter {
+
+		public bool synchronous = false;
+
+		private int POST_WRAPPER_BYTES = 88; // "schema":"iglu:com.snowplowanalytics.snowplow/payload_data/jsonschema/1-0-3","data":[]
+		private int POST_STM_BYTES = 22;     // "stm":"1443452851000",
+		private int FAIL_INTERVAL  = 10000;  // If all events failed to send
+
+		// Emitter loop variables
+		private readonly object emitLock = new object ();
+		private volatile bool sending = false;
+		private Thread emitThread;
+
+		// Emitter event store add variables
+		private ConcurrentQueue<TrackerPayload> payloadQueue = new ConcurrentQueue<TrackerPayload>();
+		private volatile bool consuming = false;
+		private Thread payloadConsumer;
+
+		// Emitter variables
+		private string endpoint;
+		private string collectorUri;
+		private HttpProtocol httpProtocol;
+		private HttpMethod httpMethod;
+		private int sendLimit;
+		private int bufferSize;
+		private long byteLimitGet;
+		private long byteLimitPost;
+		private EventStore eventStore;
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="SnowplowTracker.Emitter"/> class.
+		/// </summary>
+		/// <param name="endpoint">The collector endpoint uri</param>
+		/// <param name="protocol">What protocol to send under</param>
+		/// <param name="method">What method of sending to use</param>
+		/// <param name="sendLimit">The amount of events to pull from the database per sending attempt</param>
+		/// <param name="bufferSize">The maximum amount of events to put in a POST</param>
```

Comment:
Should we also specify the minimum number of events to put in a POST? At the moment the emitting thread waits until there is at least 1 event in the database, then tries to send every event in the database. The benefit of waiting for a larger number of events to batch is that fewer requests get made. The disadvantage is that events can languish for a long time in the database without getting sent.

Possibly we could periodically awaken the emitter thread and have it send events if either there are enough in the database or there is at least 1 and a sufficiently long time has passed since the last send. The disadvantage of this approach would be increasing complexity with yet another concurrent task.

Thoughts?